### PR TITLE
Heading: Complete TypeScript migration of component

### PR DESCRIPTION
**Pull request by walbo**
## What?
Completes the migration of `Heading` component to TypeScript.

## Why?
Part of the @wordpress/components's TypeScript migration (https://github.com/WordPress/gutenberg/issues/35744).

## How?
- Moved types to `types.js`
- Refactored test to use TypeScript
- Updated docs

## Testing Instructions
1. Run npm run dev and ensure no typing errors
2. Fire up the Storybook and confirm the `Heading` continues to function as expected
3. Test the various Headings within the block/site editors in Gutenberg still work

### CODE REVIEW

File: packages/components/src/heading/README.md [hunk 1]
```diff
@@ -20,8 +20,9 @@ function Example() {
 
 `Heading` uses `Text` underneath, so we have access to all of `Text`'s props except for `size` which is replaced by `level`. For a complete list of those props, check out [`Text`](/packages/components/src/text/README.md#props).
 
-##### level
-
-**Type**: `1 | 2 | 3 | 4 | 5 | 6`
+##### `level`: `1 | 2 | 3 | 4 | 5 | 6`
```
Comment (aaronrobertshaw) [inline packages/components/src/heading/README.md]: Nit: Given the type is for both numbers and strings, should we reflect that here for accuracy?

`export type HeadingSize = 1 | 2 | 3 | 4 | 5 | 6 | '1' | '2' | '3' | '4' | '5' | '6';`

File: packages/components/src/heading/test/index.tsx [hunk 2]
```diff
@@ -51,8 +51,8 @@ describe( 'props', () => {
 				Code is Poetry
 			</Heading>
 		);
-		expect( container.firstChild.getAttribute( 'role' ) ).toBe( 'heading' );
-		expect( container.firstChild.getAttribute( 'aria-level' ) ).toBe( '3' );
+		expect( container.firstChild ).toHaveAttribute( 'role', 'heading' );
+		expect( container.firstChild ).toHaveAttribute( 'aria-level', '3' );
```
Comment (aaronrobertshaw) [inline packages/components/src/heading/test/index.tsx]: Nit: If we are changing the previous uses of `container.firstChild` to `.children[ 0 ]` should we also do the same here?
```suggestion
		expect( container.children[ 0 ] ).toHaveAttribute( 'role', 'heading' );
		expect( container.children[ 0 ] ).toHaveAttribute( 'aria-level', '3' );
```

File: packages/components/src/heading/test/index.tsx [hunk 3]
```diff
@@ -61,7 +61,7 @@ describe( 'props', () => {
 				Code is Poetry
 			</Heading>
 		);
-		expect( container.firstChild.getAttribute( 'role' ) ).toBeNull();
-		expect( container.firstChild.getAttribute( 'aria-level' ) ).toBeNull();
+		expect( container.firstChild ).not.toHaveAttribute( 'role' );
+		expect( container.firstChild ).not.toHaveAttribute( 'aria-level' );
```
Comment (aaronrobertshaw) [inline packages/components/src/heading/test/index.tsx]: Nit: Same as above, should we be consistent with use of `.firstChild` or `.children[ 0 ]`

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for the continued efforts typing the components package @walbo 🙇

This is looking pretty good.

While testing the component in the Storybook I noticed that the `variant` control is a [radio](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/heading/stories/index.tsx#L22) button. This means you can't toggle off the `muted` variant once it's clicked. We should probably tweak that while we're here. 

Not sure if a `check` or `select` control might be better options. If using a `check` control the example markup may show `variant={[]}` when you toggle `muted` on then off. We could even switch the control to `boolean` and set the `variant` prop to `muted` based on that being set.

<details>
<summary><strong>Example diff demonstrating toggle control for variant etc.</strong></summary>

```diff
diff --git a/packages/components/src/heading/stories/index.tsx b/packages/components/src/heading/stories/index.tsx
index f25b21728e..1cc78538ac 100644
--- a/packages/components/src/heading/stories/index.tsx
+++ b/packages/components/src/heading/stories/index.tsx
@@ -19,7 +19,7 @@ const meta: ComponentMeta< typeof Heading > = {
 		letterSpacing: { control: { type: 'text' } },
 		lineHeight: { control: { type: 'text' } },
 		optimizeReadabilityFor: { control: { type: 'color' } },
-		variant: { control: { type: 'radio' }, options: [ 'muted' ] },
+		variant: { control: { type: 'boolean' } },
 		weight: { control: { type: 'text' } },
 	},
 	parameters: {
@@ -30,8 +30,9 @@ const meta: ComponentMeta< typeof Heading > = {
 export default meta;
 
 export const Default: ComponentStory< typeof Heading > = ( props ) => (
-	<Heading { ...props } />
+	<Heading { ...props } variant={ props.variant ? 'muted' : undefined } />
 );
 Default.args = {
 	children: 'Heading',
+	level: '3',
 };

```
</details>

Also, it might be beneficial to demonstrate the `level` prop's use within the default story.

Other than that, I left a couple of other very minor comments inline as well.

✅ No typing errors
✅ Unit tests pass
✅ Code changes make sense
✅ Has changelog & readme props and types pretty much match
✅ Component functions correctly in both Storybook and editors
✅ Auto-generated docs in Storybook are meaningful
❔ Storybook example has appropriate controls

File: packages/components/src/heading/test/index.tsx (hunk 2 above)
Comment (mirka) [reply packages/components/src/heading/test/index.tsx]: If anything, I'd prefer we simplify in [this direction](https://github.com/WordPress/gutenberg/compare/refactor/heading-to-typescript...refactor/heading-to-typescript-tests?expand=1#diff-a4c73bed0ef8cc8f0e9883a24338969e5f48c16aab6b97fe297b9d887f19df0d), so we don't have to reach into containers directly. I think this style better matches the Testing Library paradigm. What do folks think?

**Comment by mirka** [comment]
> While testing the component in the Storybook I noticed that the `variant` control is a [radio](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/heading/stories/index.tsx#L22) button. This means you can't toggle off the `muted` variant once it's clicked. We should probably tweak that while we're here.

I don't feel super strongly about it, but for context, here are the pros/cons that were discussed: https://github.com/WordPress/gutenberg/pull/38842#discussion_r813249434

If y'all think it's worth addressing, another way to go might be:

```js
		variant: {
			control: { type: 'radio' },
			options: [ 'undefined', 'muted' ],
			mapping: { undefined, muted: 'muted' },
		},
```

**Comment by walbo** [comment]
> > While testing the component in the Storybook I noticed that the `variant` control is a [radio](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/heading/stories/index.tsx#L22) button. This means you can't toggle off the `muted` variant once it's clicked. We should probably tweak that while we're here.
> 
> I don't feel super strongly about it, but for context, here are the pros/cons that were discussed: [#38842 (comment)](https://github.com/WordPress/gutenberg/pull/38842#discussion_r813249434)
> 
> If y'all think it's worth addressing, another way to go might be:
> 
> ```js
> 		variant: {
> 			control: { type: 'radio' },
> 			options: [ 'undefined', 'muted' ],
> 			mapping: { undefined, muted: 'muted' },
> 		},
> ```

Implemented the `undefined` option. Seem like a bettere solution since we don't have to change the templat 👍

**Review by aaronrobertshaw — APPROVED** [review]
Thanks for the updates! 

LGTM 🚢 

✅ No typing errors
✅ Unit tests pass
✅ Code changes make sense
✅ Has changelog & readme props and types match
✅ Component functions correctly in both Storybook and editors
✅ Auto-generated docs in Storybook are meaningful
✅ Storybook example has appropriate controls

**Pull request merged**